src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,332 @@
+        /// </summary>
+        private const int DefaultMaxNotValidBeforeDelta = 140;
+        /// <summary>
+        /// A default value for deposit lock period.
+        /// </summary>
+        private const int DefaultDepositDeltaTill = 5760;
+        private const byte Prefix_Deposit = 1;
+        private const byte Prefix_MaxNotValidBeforeDelta = 10;
+
+        internal Notary() : base() { }
+
+        public override Hardfork? ActiveIn => Hardfork.HF_Echidna;
+
+        internal override ContractTask InitializeAsync(ApplicationEngine engine, Hardfork? hardfork)
+        {
+            if (hardfork == ActiveIn)
+            {
+                engine.SnapshotCache.Add(CreateStorageKey(Prefix_MaxNotValidBeforeDelta), new StorageItem(DefaultMaxNotValidBeforeDelta));
+            }
+            return ContractTask.CompletedTask;
+        }
+
+        internal override async ContractTask OnPersistAsync(ApplicationEngine engine)
+        {
+            long nFees = 0;
+            ECPoint[] notaries = null;
+            foreach (Transaction tx in engine.PersistingBlock.Transactions)
+            {
+                var attr = tx.GetAttribute<NotaryAssisted>();
+                if (attr is not null)
+                {
+                    notaries ??= GetNotaryNodes(engine.SnapshotCache);
+                    var nKeys = attr.NKeys;
+                    nFees += (long)nKeys + 1;
+                    if (tx.Sender == Hash)
+                    {
+                        var payer = tx.Signers[1];
+                        var balance = engine.SnapshotCache.GetAndChange(CreateStorageKey(Prefix_Deposit, payer.Account))?.GetInteroperable<Deposit>();
+                        balance.Amount -= tx.SystemFee + tx.NetworkFee;
+                        if (balance.Amount.Sign == 0) RemoveDepositFor(engine.SnapshotCache, payer.Account);
+                    }
+                }
+            }
+            if (nFees == 0) return;
+            var singleReward = CalculateNotaryReward(engine.SnapshotCache, nFees, notaries.Length);
+
+            foreach (var notary in notaries) await GAS.Mint(engine, notary.EncodePoint(true).ToScriptHash(), singleReward, false);
+        }
+
+        protected override void OnManifestCompose(IsHardforkEnabledDelegate hfChecker, uint blockHeight, ContractManifest manifest)
+        {
+            manifest.SupportedStandards = new[] { "NEP-27" };
+        }
+
+        /// <summary>
+        /// Verify checks whether the transaction is signed by one of the notaries and
+        /// ensures whether deposited amount of GAS is enough to pay the actual sender's fee.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="sig">Signature</param>
+        /// <returns>Whether transaction is valid.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        private bool Verify(ApplicationEngine engine, byte[] sig)
```

Comment:
Validate sig argument? != null and length?

Reply:
Fixed.